### PR TITLE
Addressing Emission results conversion and region hard coded extraction

### LINKED ISSUE #24: openentrance dependency must be installed manually
- [x] Add description to circumvent issue to the readme
- [ ] Monitor master issue openENTRANCE/openentrance#202

### PR DESCRIPTION
**Pull request by genius487**
Good day. Here is a summary of my changes.

- removed the need for openentrance package (closes #24)
	+ custom dictionary module was created to convert any country's alpha-2 code to the ISO english short names
		+ the function can also display an (error) message shows if the first 2 letters weren't country codes but doesn't the code is still allowed to run. The code just completes, returning what was there

- changed read_file function to be able to read results generated from KTH's interface
	+ The region is determined when the file is read per result from the config file, depending on what the naming option is as written in the config file. ```from-file``` uses the region in the CSV. ```from-tech-name``` uses the region from the ISO codes at the beginning of technology and fuel names (for OSeMBE purposes mainly). Otherwise, whatever country name is written in the config file is forced as the region.

My config that I used testing is [here](https://github.com/OSeMOSYS/osemosys2iamc/files/11814789/testing-config.txt)

### CODE REVIEW

**Review by willu47 — CHANGES_REQUESTED** [review]
Hi @genius487. Many thanks for your contribution!

It is a good idea to remove the openentrance package, particularly as it was only used to obtain the ISO2 country codes.  However, I think there could be an easier way than hardcoding the data into the code.  For example, the country codes are available on [datahub](https://datahub.io/core/country-list#python).  There's also [a Python package](https://pypi.org/project/iso3166/) which provides access to the country codes.

The `read_file()` function you've provided is going to be very fragile, as different OSeMOSYS result files do not have the same index/columns. This information is available in the config.yaml file which the user may have when using [otoole](https://github.com/OSeMOSYS/otoole) to convert the results.  Also, this does not solve the issue of hard-coding region name extraction.  Not all models use the first two letters of a technology or fuel name to denote a region.

To resolve #27, the user should specify in the config file how region names are encoded.  The options could be:

- country names: iso2, iso3, full name
- position: region index, technology name & position e.g. 0:2; fuel name & position;

File: src/osemosys2iamc/resultify.py [hunk 1]
```diff
@@ -26,9 +26,40 @@
...
 
 
-def read_file(filename) -> pd.DataFrame:
+def read_file(in_path: str, osemosys_param: str, region_name_option: str) -> pd.DataFrame:
+    """Reads in selected csv file as given in the config file into a Pandas DataFrame
 
+    Also, at the time of DataFrame creation, this function checks how it should derive
+    the name of the region to be used
+
+    Options:
```
Comment (willu47) [inline src/osemosys2iamc/resultify.py:35]: Please use the docstrings standard implemented in the rest of the package.  Replace `options:` with 
```python
Arguments
---------
```
Also please use underscores and not hyphens for your argument names.

File: src/osemosys2iamc/resultify.py [hunk 2]
```diff
@@ -26,9 +26,40 @@
...
+        from-tech-name: Converts the first two characters of the technology/fuel to
```
Comment (willu47) [inline src/osemosys2iamc/resultify.py:36]: This does not describe the arguments in the function

File: src/osemosys2iamc/resultify.py [hunk 3]
```diff
@@ -26,9 +26,40 @@
...
+def read_file(in_path: str, osemosys_param: str, region_name_option: str) -> pd.DataFrame:
+    """Reads in selected csv file as given in the config file into a Pandas DataFrame
 
+    Also, at the time of DataFrame creation, this function checks how it should derive
+    the name of the region to be used
+
+    Options:
+        from-tech-name: Converts the first two characters of the technology/fuel to 
+                        the country's name
+        from-file: Uses the region specified in the CSV file
```
Comment (willu47) [inline src/osemosys2iamc/resultify.py:38]: `from-file` is not an argument

File: src/osemosys2iamc/resultify.py [hunk 4]
```diff
@@ -26,9 +26,40 @@
 import matplotlib.dates as mdates
 
 
-def read_file(filename) -> pd.DataFrame:
+def read_file(in_path: str, osemosys_param: str, region_name_option: str) -> pd.DataFrame:
```
Comment (willu47) [inline src/osemosys2iamc/resultify.py:29]: This function reads results from a different result file format to that used by other models. If this is a requirement, I suggest first writing a script which converts all the results to the standard csv format defined by otoole, and then using the existing code.

**Review by willu47 — CHANGES_REQUESTED** [review]

**Comment by genius487** [comment]
@willu47 Thanks so much for the feedback. I amended the workflow by using the python package you suggested. Also, this is now a partial solution of the region hard coding as the iso2 or iso3 codes can be located anywhere in the technology or fuel name and converted to the full name of the country. Kindly inform if this is satisfactory.

File: src/osemosys2iamc/resultify.py [hunk 5]
```diff
@@ -123,14 +216,11 @@ def filter_capacity(df: pd.DataFrame, technologies: List[str]) -> pd.DataFrame:
 def filter_final_energy(df: pd.DataFrame, fuels: List) -> pd.DataFrame:
     """Return dataframe that indicate the final energy demand/use per country and year.
     """
-    for f in fuels:
-        if len(f)!=2:
-            print("Fuel %s from config.yaml doesn't comply with expected format." % f)
-            exit(1)
-
-    df['REGION'] = df['FUEL'].str[:2]
+    #for f in fuels:
```
Comment (willu47) [inline src/osemosys2iamc/resultify.py:219]: You can delete code that is not needed.

**Comment by willu47** [comment]
> @willu47 Thanks so much for the feedback. I amended the workflow by using the Python package you suggested. Also, this is now a partial solution of the region hard coding as the iso2 or iso3 codes can be located anywhere in the technology or fuel name and converted to the full name of the country. Kindly inform if this is satisfactory.

This looks good.

I think you should add some tests for the iso2country method.  Take a look in the `tests/` folder and create a new test function in `test_resultify.py`.

I would be interested in a few corner cases such as:

NGNGA2 - natural gas technology in Nigeria (iso2)
DENGA2 - natural gas technology in Germany
NGKENGX - export of natural gas from Nigeria to Kenya
NGANGA - natural gas technology in Nigeria (iso3)
ZXNGA - country code does not exist natural gas technology
'strat', 'ned' - misspelt arguments should raise a ValueError()
-1, 22 - position number negative or larger than string should raise a ValueError()

You can run the tests with pytest

**Comment by genius487** [comment]
@willu47 I've added some tests, and renamed the function to iso_to_country for deeper clarity. A position larger than the string's length isn't intended to raise an error in order to accommodate a list of name with varying name length. Instead, an empty string is returned and a message notifying the failure to find the code is displayed for the user.

**Review by willu47 — APPROVED** [review]
